### PR TITLE
[Ability] Implement Ability Illusion

### PR DESCRIPTION
**Pull request by PyGaVS**
## What are the changes?

Adding illusion ability, works with fusion also

## Why am I doing these changes?

It's missing

## What did change?

-Added PreSummonAbAttr class as ability attribute.
-Illusion uses the appearance of the last conscious and available Pokémon in the party (ignoring itself), including :
Species, Form, Shininess, Pokeball, Gender
-If We encounter a wild Zorua/Zoroak he will use the illusion of a pokemon in the Abyss Biome. Depending on Zorua or Zoroark level the illusion will be evolved or not.
-If we encounter a boss Zoroark he will use the illusion of Entei, Suicune or Raiku like in the original game.
-The class pokemon has a new attribute illusion.
-A pokemon with illusion can use illusion one time per battle like in the original game.
-After a battle, the pokemon can use illusion again.
-The AI move choice will be affected by the illusion.
-The type will always be displayed as the Pokémon it is disguised as, even when affected by Soak etc.
-The illusion will disappear if the Pokemon has its ability changed or suppressed.
-Pokemon cannot transform into a Pokémon that is disguised by Illusion.
-Pokémon that are disguised by Illusion cannot Transform.
-Illusion will not activate if the last conscious Pokémon is a Terastallized Ogerpon or Terapagos.

### Screenshots/Videos

https://discord.com/channels/1125469663833370665/1176874654015684739/1257835032647831682
https://discord.com/channels/1125469663833370665/1176874654015684739/1257864381937619045
https://discord.com/channels/1125469663833370665/1176874654015684739/1257865512801538059
https://discord.com/channels/1125469663833370665/1176874654015684739/1258418548074745958

## How to test the changes?

I played Zoroark and override the enemy pokemon with Zoroark so we can see him using correct illusion
I used illusion on a fusion and with a fusion
I tried to switch with the illusion
We can also run `npm run test illusion`

Old conversations here : 
https://github.com/pagefaultgames/pokerogue/pull/2217

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: package-lock.json
Comment (flx-sta) [inline package-lock.json:1]: Why was `package-lock.json` changed? This should only occur when necessary (e.g. dependency update)

File: public/images/items.json
Comment (flx-sta) [inline public/images/items.json:1]: why is this considered 'changed' in this PR?

File: src/data/ability.ts [hunk 1]
```diff
@@ -4041,6 +4049,97 @@ export class IceFaceBlockPhysicalAbAttr extends ReceivedMoveDamageMultiplierAbAt
...
+   * @param {Pokemon} attacker - The attacking Pokémon.
+   * @param {PokemonMove} move - The move being used.
+   * @param {PokemonMove} hitResult - The type of hitResult the pokemon got
+   * @param {...any} args - N/A
+   * @returns {boolean} - Whether the illusion was destroyed.
+   */
+  applyPostDefend(pokemon: Pokemon, passive: boolean, attacker: Pokemon, move: Move, hitResult: HitResult, args: any[]): boolean {
+
+    const breakIllusion: HitResult[] = [HitResult.EFFECTIVE, HitResult.SUPER_EFFECTIVE, HitResult.NOT_VERY_EFFECTIVE, HitResult.ONE_HIT_KO];
+    if (!breakIllusion.includes(hitResult)) {
```
Comment (flx-sta) [inline src/data/ability.ts:4101]: 🚀 better 👍

File: src/data/ability.ts [hunk 2]
```diff
@@ -4272,7 +4376,7 @@ export function applyCheckTrappedAbAttrs(attrType: Constructor<CheckTrappedAbAtt
   return applyAbAttrsInternal<CheckTrappedAbAttr>(attrType, pokemon, (attr, passive) => attr.applyCheckTrapped(pokemon, passive, trapped, otherPokemon, args), args);
 }
 
-export function applyPostBattleAbAttrs(attrType: Constructor<PostBattleAbAttr>,
+export function applyPostBattleAbAttrs(attrType: { new(...args: any[]): PostBattleAbAttr },
```
Comment (flx-sta) [inline src/data/ability.ts:4379]: What's the reason for breaking `Constructor<PostBattleAbAttr>` here?

File: public/images/items.json
Comment (PyGaVS) [reply public/images/items.json:1]: i don't know i merged beta branch and it's still here, i should maybe copy the last character manually.

File: src/data/ability.ts (hunk 2 above)
Comment (PyGaVS) [reply src/data/ability.ts:4379]: it was not here when i implemented that i will apply corrections

File: public/images/items.json
Comment (flx-sta) [reply public/images/items.json:1]: if you change to hide `whitespace changes`
image


you can see that it has to do with it

File: package-lock.json
Comment (PyGaVS) [reply package-lock.json:1]: it's not changed

File: public/images/items.json
Comment (PyGaVS) [reply public/images/items.json:1]: ok thanks but i still don't know how to solve this lol

**Pull request closed without merging**